AX: After ENABLE(AX_THREAD_TEXT_APIS), VoiceOver speech becomes out of sync with the cursor position when navigating by line when starting at the start or end of soft line break positions

twilco opened the pull request:
#### e4deb1dd40c5010b6b0e986ae0a2de1876947e9a
<pre>
AX: After ENABLE(AX_THREAD_TEXT_APIS), VoiceOver speech becomes out of sync with the cursor position when navigating by line when starting at the start or end of soft line break positions
<a href="https://bugs.webkit.org/show_bug.cgi?id=291437">https://bugs.webkit.org/show_bug.cgi?id=291437</a>
<a href="https://rdar.apple.com/149081361">rdar://149081361</a>

Reviewed by Joshua Hoffman.

This happened because we were not properly considering affinity, which is important at soft line break boundaries.
Fix the issue by making sure we keep affinity given to us when created text markers from VisiblePositions, and setting
affinity ourselves when appropriate in AXTextMarker::findLine.

Then we take affinity into account in AXTextRuns::indexForOffset, as what text run (and thus what line) a marker belongs
to is a function of both offset *and* affinity.

New tests perfectly simulate the issue, and one can reproduce the bug in the browser by issuing the same keystrokes
laid out in the test.

This also happened to fix existing tests mac/attributed-string-with-listitem-multiple-lines.html and mac/line-index-for-textmarker.html.

* LayoutTests/accessibility-isolated-tree/TestExpectations:
* LayoutTests/accessibility/mac/editable-single-letter-soft-linebreak-lines-expected.txt: Added.
* LayoutTests/accessibility/mac/editable-single-letter-soft-linebreak-lines.html: Added.
* LayoutTests/accessibility/mac/line-range-in-contenteditable-expected.txt: Added.
* LayoutTests/accessibility/mac/line-range-in-contenteditable.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::textMarkerDataForVisiblePosition):
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::clampOffsetToLengthIfNeeded const):
(WebCore::AXTextMarker::markerRangeForLineIndex const):
(WebCore::AXTextMarker::lineNumberForIndex const):
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::findLine const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::affinity const): Added.
(WebCore::AXTextMarker::isDownstream const): Added.
(WebCore::AXTextMarker::setAffinity):
* Source/WebCore/accessibility/AXTextRun.cpp:
(WebCore::AXTextRuns::indexForOffset const):
(WebCore::AXTextRuns::localRect const):
(WebCore::AXTextRuns::lineIDForOffset const):
Remove unused function.
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRuns::lastRunIndex const):

Canonical link: <a href="https://commits.webkit.org/293733@main">https://commits.webkit.org/293733@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36380bbc09597681df9830ff599814e91b7ad226

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99731 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19379 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9650 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104861 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50326 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101772 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19683 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27814 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75923 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33016 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102738 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14999 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90066 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56283 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14802 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8053 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49686 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8138 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107221 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26845 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19608 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84877 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27209 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86271 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84396 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21436 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29077 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6794 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20665 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26786 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31994 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26600 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29915 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28169 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -1103,12 +1106,44 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, CoalesceObjectBreak
+        // Example, where _ is a whitespace:
+        // foo_
+        // |bar
+        return runIndex;
```
Optional, but I'd favor return runIndex > 0; so it's clear you're returning a bool

twilco replied:
Agreed, fixed!

minorninth commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -1103,12 +1106,44 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, CoalesceObjectBreak
     if (!isInTextRun())
         return toTextRunMarker(stopAtID).findLine(direction, boundary, includeTrailingLineBreak, stopAtID);
+    // Temporarily change our offset for the purposes of determining line positioning if at an ambiguous position.
+    auto adjustedOffsetScope = SetForScope(const_cast<AXTextMarker*>(this)->m_data.offset, shouldIncrementOffsetForAmbiguousPosition() ? m_data.offset + 1 : m_data.offset);
```
Is incrementing the offset always safe? What if the downstream line is empty and offsetting by 1 moves you down one more line?

hoffmanjoshua commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -784,6 +785,8 @@ bool AXTextMarker::atLineBoundaryForDirection(AXDirection direction) const
     if (!isInTextRun())
         return toTextRunMarker().atLineBoundaryForDirection(direction);
+    // Temporarily change our offset for the purposes of determining line positioning if at an ambiguous position.
+    auto adjustedOffsetScope = SetForScope(const_cast<AXTextMarker*>(this)->m_data.offset, shouldIncrementOffsetForAmbiguousPosition() ? m_data.offset + 1 : m_data.offset);
```
I notice we are doing this in this overload of `atLineBoundaryForDirection`. Is it going to cause any issues for callsities that use the three parameter overload?

twilco replied:
The only callers of the three-parameter overload are `findLine` and the single-parameter version of the function, so I think we're technically OK. But if someone started added a new direct use of the three-parameter overload, that would indeed be problematic 🤔 

I could add something like this to the three-parameter overload:

```cpp
// Any ambiguous positions should've been fixed by the time this function was called. If not, we're going to behave incorrectly.
ASSERT(!shouldIncrementOffsetForAmbiguousPosition());
```

But open to other ideas if ya have any.

hoffmanjoshua replied:
I think that's a great catchall solution! I don't anticipate this is the last place we'll add the `atLineBoundaryForDirection` family of methods, so better to be safe now.

Can we use one of the text marker assert macros?

minorninth reviewed (approved):
lgtm. Passing Affinity AXTextRuns::indexForOffset makes a lot of sense. Thanks!

hoffmanjoshua commented on Source/WebCore/accessibility/AXTextRun.h:
```diff
@@ -153,8 +155,7 @@ struct AXTextRuns {
 
-    size_t indexForOffset(unsigned textOffset) const;
-    AXTextRunLineID lineIDForOffset(unsigned textOffset) const;
+    size_t indexForOffset(unsigned textOffset, Affinity) const;
```
nit: should we consider having a default affinity here for future uses?

```cpp
size_t indexForOffset(unsigned textOffset, Affinity = Affinity::Upstream) const;
```

twilco replied:
I think we want affinity to be explicit since it's important, wouldn't want us to accidentally forget to pass it and get the wrong behavior 🤔

hoffmanjoshua replied:
Makes sense—don’t worry about this then

twilco committed:
AX: After ENABLE(AX_THREAD_TEXT_APIS), VoiceOver speech becomes out of sync with the cursor position when navigating by line when starting at the start or end of soft line break positions
https://bugs.webkit.org/show_bug.cgi?id=291437
rdar://149081361

Reviewed by Joshua Hoffman.

This happened because we were not properly considering affinity, which is important at soft line break boundaries.
Fix the issue by making sure we keep affinity given to us when created text markers from VisiblePositions, and setting
affinity ourselves when appropriate in AXTextMarker::findLine.

Then we take affinity into account in AXTextRuns::indexForOffset, as what text run (and thus what line) a marker belongs
to is a function of both offset *and* affinity.

New tests perfectly simulate the issue, and one can reproduce the bug in the browser by issuing the same keystrokes
laid out in the test.

This also happened to fix existing tests mac/attributed-string-with-listitem-multiple-lines.html and mac/line-index-for-textmarker.html.

* LayoutTests/accessibility-isolated-tree/TestExpectations:
* LayoutTests/accessibility/mac/editable-single-letter-soft-linebreak-lines-expected.txt: Added.
* LayoutTests/accessibility/mac/editable-single-letter-soft-linebreak-lines.html: Added.
* LayoutTests/accessibility/mac/line-range-in-contenteditable-expected.txt: Added.
* LayoutTests/accessibility/mac/line-range-in-contenteditable.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::textMarkerDataForVisiblePosition):
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::clampOffsetToLengthIfNeeded const):
(WebCore::AXTextMarker::markerRangeForLineIndex const):
(WebCore::AXTextMarker::lineNumberForIndex const):
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::findLine const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::affinity const): Added.
(WebCore::AXTextMarker::isDownstream const): Added.
(WebCore::AXTextMarker::setAffinity):
* Source/WebCore/accessibility/AXTextRun.cpp:
(WebCore::AXTextRuns::indexForOffset const):
(WebCore::AXTextRuns::localRect const):
(WebCore::AXTextRuns::lineIDForOffset const):
Remove unused function.
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRuns::lastRunIndex const):

Canonical link: https://commits.webkit.org/293733@main
